apic_ml2/neutron/services/l3_router/apic_driver.py

```diff
@@ -0,0 +1,246 @@
+        if 'subnet_id' in interface_info:
+            subnet = self._plugin._core_plugin.get_subnet(
+                context, interface_info['subnet_id'])
+            network_id = subnet['network_id']
+            port_id = self._get_port_id_for_router_interface(
+                context, router_id, interface_info['subnet_id'])
+        else:
+            port = self._plugin._core_plugin.get_port(
+                context, interface_info['port_id'])
+            network_id = port['network_id']
+            port_id = interface_info['port_id']
+
+        network = self._plugin._core_plugin.get_network(context, network_id)
+        tenant_id = network['tenant_id']
+
+        router = self._plugin.get_router(context, router_id)
+        # Map openstack IDs to APIC IDs
+        atenant_id, arouter_id, anetwork_id, _ = self._map_names(
+            context, tenant_id, router, network, None)
+
+        # Program APIC
+        self.manager.remove_router_interface(
+            self.aci_mech_driver._get_network_aci_tenant(network),
+            arouter_id, anetwork_id,
+            app_profile_name=self.aci_mech_driver._get_network_app_profile(
+                network))
+        self._plugin._core_plugin.update_port_status(
+            context, port_id, q_const.PORT_STATUS_DOWN)
+
+    def delete_router_precommit(self, context, router_id):
+        context._plugin = self
+        router = self._plugin.get_router(context, router_id)
+        with apic_mapper.mapper_context(context) as ctx:
+            arouter_id = router_id and self.name_mapper.router(
+                ctx, router['id'], openstack_owner=router['id'])
+        self.manager.delete_router(arouter_id)
+
+    def update_router_postcommit(self, context, router):
+        context._plugin = self
+        with apic_mapper.mapper_context(context) as ctx:
+            arouter_id = router['id'] and self.name_mapper.router(
+                ctx, router['id'], openstack_owner=router['tenant_id'])
+            tenant_id = self.aci_mech_driver._get_router_aci_tenant(router)
+
+        with self.manager.apic.transaction() as trs:
+            vrf = self.aci_mech_driver._get_tenant_vrf(router['tenant_id'])
+            # A Neutron router is rendered as a contract. This contract
+            # Will exist in the COMMON tenant when single tenant mode is false,
+            # in the sys_id tenant otherwise.
+            self.manager.create_router(arouter_id, owner=tenant_id,
+                                       transaction=trs,
+                                       context=vrf['aci_name'])
+            if router['admin_state_up']:
+                self.manager.enable_router(arouter_id, owner=tenant_id,
+                                           transaction=trs)
+            else:
+                self.manager.disable_router(arouter_id, owner=tenant_id,
+                                            transaction=trs)
+
+    def create_floatingip_postcommit(self, context, fip_context):
```

Comment:
If I understand correctly -- we add members to the context object dynamically, so that state can be passed between the pre- and post- calls?  I say that only because we don't really control the context objects (they're part of the core Cisco Router Service Plugin code base). If that's a reasonable approach, then I'll get rid of the fip_context objects and use the base context objects instead.